### PR TITLE
Allow committing text to trigger EditableText.onChanged

### LINKED ISSUE #128565: isComposingRangeValid IOS not working
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. flutter 3.10.1
2. Dart 3.0.1

### Expected results

In the old version, when I input Chinese pinyin, it will be triggered
The new version does not work

### Actual results

In the old version, when I input Chinese pinyin, it will be triggered
The new version does not work

### Code sample

<details><summary>Code sample</summary>

```dart
[Paste your code here]
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[Paste your output here]
```

</details>

### PR DESCRIPTION
**Pull request by LinXunFeng**
Fixes [#128565](https://github.com/flutter/flutter/issues/128565)

If using Android native project to verify, we will find that the issue [#111651](https://github.com/flutter/flutter/issues/111651) is not a bug, and it is the same under iOS.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -3616,17 +3616,15 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
-        widget.onChanged?.call(currentText);
-      } catch (exception, stack) {
-        FlutterError.reportError(FlutterErrorDetails(
-          exception: exception,
-          stack: stack,
-          library: 'widgets',
-          context: ErrorDescription('while calling onChanged'),
-        ));
-      }
+    try {
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:3619]: Should this check if `.text` OR `.composing` are different and only call onChanged then?

**Review by justinmc — COMMENTED** [review]
Can you help me understand exactly what problem you're trying to fix with this PR?  I'm confused how the issues you mention in the description exactly relate to this.

Changing the behavior of onChanged will likely be a breaking change so I want to make sure we understand what the correct behavior should be here before we commit to changing it.

Would adding something like `onComposingChanged` as a separate method also help your situation?  Just thinking out loud.

**Comment by LinXunFeng** [comment]
The problem that typing `a111` will print twice mentioned in issue [#111651](https://github.com/flutter/flutter/issues/111651) was mistaken for a bug, and it was  "fixed" in PR [#112010](https://github.com/flutter/flutter/pull/112010).

We can use the native project to see the correct behavior.

### Android
link: [native_android_project](https://github.com/LinXunFeng/native_textfield_demo/tree/main/android_project)

```kotlin
val etTest = findViewById<EditText>(R.id.etTest)
etTest.addTextChangedListener(object : TextWatcher {
    override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
        log("beforeTextChanged: ${s}, start = ${start}, count = ${count}, after = ${after}")
    }
    override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
        log("onTextChanged: ${s}, start = ${start}, before = ${before}, count = ${count}")
    }
    override fun afterTextChanged(s: Editable?) {
        log("afterTextChanged: ${s?.toString()}")
        log("==================================")
    }
})
```

image

When typing to `a111`, the `afterTextChanged` method will be called twice.

We can learn from the logic of `isComposingRangeValid` in Flutter to write an `isComposingRangeValid` for Android native, here it is named `isComposition`

```kotlin
override fun afterTextChanged(s: Editable?) {
    val isComposition = isComposition(s)
    if (isComposition) {
        log("isComposition: ${s?.toString()}")
    } else {
        log("text: ${s?.toString()}")
    }
}
```
We can use it to determine whether it is the finalized text. 
> ps: Unfinalized text will be underlined.

https://github.com/flutter/flutter/assets/19367531/4b104844-06ec-41e8-b155-0b53bef24055

### iOS
link: [native_ios_project](https://github.com/LinXunFeng/native_textfield_demo/tree/main/ios_project)

```objc
- (void)viewDidLoad {
  ...

  // Listen text input
  [self.textField addTarget:self
                     action:@selector(textFieldDidChange:)
           forControlEvents:UIControlEventEditingChanged];
}

- (void)textFieldDidChange:(UITextField *)textField {
    if (textField.markedTextRange != nil) {
      // Did not click the Chinese characters on accessory view, keep typing.
      NSLog(@"markedTextRange: %@",textField.text);
    } else {
      // Commit text, which means that the text has been determined.
      // This is where you do things like make network requests or filter data.
      NSLog(@"text: %@", textField.text);
    }
}
```

https://github.com/flutter/flutter/assets/19367531/5241a291-1edc-44c9-96b4-b00220c9dbe2

As can be seen from the above video, when I click the keyboard, the `textFieldDidChange` method will be called, but at this time the text is unfinalized, which can be judged by `markedTextRange != nil`, when I click the recommended word "大家好" , the `textFieldDidChange` method  is called twice, at this time we can judge whether the text is finalized by `markedTextRange == nil`.

So, the `textFieldDidChange` method is equivalent to the `onChanged` callback in Flutter, and `markedTextRange == nil` is equivalent to `isComposingRangeValid` in Flutter.

From the above content, I think it is normal for the `onchange` callback in Flutter to be triggered multiple times. We can write code like this in the `onChanged` callback

```dart
TextField(
  controller: textController,
  onChanged: (value) {
    if (textController.value.isComposingRangeValid) {
      return;
    }
    print('make a network request or filter data');
  },
),
```

The change of this PR is only to revert to the behavior of  Flutter 3.3.x and before, it is not a breaking change .

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 1 above)
Comment (LinXunFeng) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Sorry, I thought the above code was judged, I have adjusted the code now.

**Comment by LongCatIsLooong** [comment]
This change may break https://github.com/flutter/flutter/issues/111651 again. Historically the `onChanged` callback is only called when the text changes (e.g.: https://github.com/flutter/flutter/blob/5391447fae6209bb21a89e6a5a6583cac1af9b4b/packages/flutter/lib/src/widgets/editable_text.dart#L821-L823) and I think it's better keep it that way to avoid potential breakage. The API doesn't have to behave exactly the same as Android `EditText` text change listeners or what `UIKit` `UITextField` offers. 

However I do think it would be nice to have a different `onChanged` callback that provides more information.

**Pull request closed without merging**